descqa/NumberDensityVersusRedshift.py

```diff
@@ -0,0 +1,307 @@
+        #make sure data entries are all finite 
+        data = GCRQuery(*((np.isfinite, col) for col in data)).filter(data)
+
+        return data
+
+    def get_validation_data(self, band, observation):
+
+        validation_data = {}
+        z0errors = np.asarray([])
+        datafile = data_info[observation]['filename_template'].format(band)
+        filename = os.path.join(self.data_dir,datafile)
+        usecols = data_info[observation]['usecols']
+        skiprows= data_info[observation]['skiprows']
+
+        if os.path.exists(filename) and os.path.getsize(filename) > 0:
+            if observation.find(Coil2004)!=-1:
+                if len(usecols)==4:
+                    mag_hi,mag_lo,z0values,z0errors = np.loadtxt(filename,unpack=True,usecols=usecols,skiprows=skiprows)
+                    validation_data['z0errors'] = z0errors
+                else:
+                    mag_hi,mag_lo,z0values = np.loadtxt(filename,unpack=True,usecols=usecols,skiprows=skiprows)
+                validation_data['z0values'] = z0values
+                validation_data['mag_hi'] = mag_hi
+                validation_data['mag_lo'] = mag_lo
+                validation_data['label'] = Coil2004_lbl
+
+            elif observation.find(DEEP2_JAN)!=-1:
+                mag_hi_lim,mag_lo_lim,z0const,z0linear = np.loadtxt(filename,unpack=True,usecols=[0,1,2,3],skiprows=1)
+                validation_data['z0const'] = z0const
+                validation_data['z0linear'] = z0linear
+                validation_data['mag_hi'] = []
+                validation_data['mag_lo'] = [float(m) for m in range(int(mag_hi_lim),int(mag_lo_lim)+1)]
+                validation_data['label'] = DEEP2_JAN_lbl
+            else:
+                raise ValueError('Observation {} not available'.format(observation))
+        else:
+            raise ValueError("{}-band data file {} not found or is empty".format(band,filename))
+            
+        return validation_data
+
+    def run_on_single_catalog(self, galaxy_catalog, catalog_name, base_output_dir):
+        
+        #check for skip_messages
+        if(hasattr(self,'skip_message')):
+            return TestResult(skipped=True, summary= self.skip_message)
+
+        #get catalog data
+        mag_field = galaxy_catalog.first_available(*self.possible_mag_fields)
+        if not mag_field:
+            return TestResult(skipped = True,summary = 'Missing requested quantities')
+        catalog_data = self.get_catalog_data(galaxy_catalog, [self.z, mag_field], filters=self.filters)
+        filtername = mag_field.partition(self.band + '_')[-1].upper()
+        filelabel = '_'.join([filtername,self.band])
+
+        fig, ax  =  plt.subplots(self.nrows,self.ncolumns,figsize=(figx_p,figy_p),sharex='col')
+        self.yaxis = 'P(z|m)' if self.normed else 'N(z|m)'
+
+        catalog_color = next(self.colors)
+        
+        #loop over magnitude cuts and make plots
+        results = {}
+        for n,cut in enumerate(self.mag_lo):
```

Comment:
Line 171 to 190
```python
for n, (ax_this, summary_ax_this, cut_lo, cut_hi, z0, z0err) in enumerate(zip_longest(
    self.ax.flat,
    self.summary_ax.flat,
    self.validation_data['mag_lo'],
    self.validation_data.get('mag_hi', []),
    self.validation_data.get('z0values', []),
    self.validation_data.get('z0errors', []),
)):

    mask = (catalog_data[mag_field] < cut_lo)
    cutlabel = '{} < {}'.format(self.band, cut_lo)
    if cut_hi:
        mask &= (catalog_data[mag_field] >= cut_hi)
        cutlabel = '{} \\leq'.format(cut_hi) + cutlabel
    cutlabel = '${}$'.format(cutlabel)

    if z0 is None:
        z0 = self.validation_data['z0const'] + self.validation_data['z0linear'] * cut_lo

    z_this = catalog_data[self.z][mask]
    del mask
    total = '(# of galaxies = {})'.format(len(z_this))
```

Reply:
Done, but changed to 
            if z0 is None and self.validation_data.has_key('z0const'):
                z0 = self.validation_data['z0const'] + self.validation_data['z0linear'] * cut_lo
to cover the case when no validation data is available. (z0 can be None if 'z0values' is not a key or validation_data is {})